tensorflow/python/kernel_tests/xent_op_deterministic_test.py

```diff
@@ -94,39 +73,132 @@ def _generateInputs(self, dtype, seed=123):
 
   @test_util.run_in_graph_and_eager_modes
   def testForward(self):
-    with self.session(), test_util.force_cpu():
-      for dtype in [np.float16, np.float32, np.float64]:
-        for trial in range(5):
-          seed = 123 + trial
-          labels, logits = self._generateInputs(dtype, seed=seed)
-          result_a = nn_ops.softmax_cross_entropy_with_logits_v2(
-              labels=labels, logits=logits)
-          result_b = nn_ops.softmax_cross_entropy_with_logits_v2(
-              labels=labels, logits=logits)
-          self.assertAllEqual(result_a, result_b)
+    for use_gpu in [False, True]:
+      if use_gpu and self._gpuNotAvailable():
+        continue
+      with self.cached_session(), test_util.device(use_gpu):
+        # Even in eager mode, the above line will be able to pin ops to CPU.
+        for dtype in [np.float16, np.float32, np.float64]:
+          for trial in range(5):
+            seed = 123 + trial
+            labels, logits = self._generateInputs(
+                dtype, seed=seed, forward_not_backward=True)
+            result_a = nn_ops.softmax_cross_entropy_with_logits(
+                labels=labels, logits=logits)
+            result_b = nn_ops.softmax_cross_entropy_with_logits(
+                labels=labels, logits=logits)
+            self.assertAllEqual(result_a, result_b)
+
+  def _testBackward(self, labels_not_logits=False):
+    for use_gpu in [False, True]:
```

Comment:
Our CI runs tests both with and without a GPU, so explicitly running on the CPU is not necessary.

Reply:
Are you saying that because every test will get run on a machine with a GPU and also on a machine without a GPU: (1) if we want a test to be run on both a CPU and a GPU then we would (in a test that will run in eager mode) use `test_util.device(use_gpu=True)` (or, do nothing); (2) if we want it to be run on the CPU (and not the GPU) in both machines (because, say, the test won't run on GPU) then we would use `test_util.device(use_gpu=False)`; (3) if we want the test to fail if we try to run it on a machine with no GPU, then we would use `test_util.force_gpu()`.

The question I have is: how do I develop and use these tests locally without having to add and remove the ability for them to be run on the different devices. In other words, I can't then as confidently submit what I have confirmed to be good code. I guess I can set `CUDA_VISIBLE_DEVICES=-1` and run the test that way to check the non-GPU testing functionality. The problem with this is that I then have to pass the environment variable in on the bazel build, which is going to lead to a complete rebuild. So it would make development take forever. Alternatively, I would need two development machines and then I would need access to the code I'm developing from both machines at once (e.g. through a mounted drive). I already have to do a rebuild to check for the XLA case, but I usually do that at the very end of development and I'm usually not particularly developing to run on XLA (it just has to not fail).

I see a lot of tests that explicitly test on both CPU and GPU, and I written a lot of tests like this myself. In the above code, I have tried to optimize a little so that only the CPU case gets run on a machine with no GPU. Many tests, including the existing xent opt tests, will run the same case twice on a machine with no GPU.

Reply:
Are you saying that because every test will get run on a machine with a GPU and also on a machine without a GPU: (1) if we want a test to be run on both a CPU and a GPU then we would (in a test that will run in eager mode) use test_util.device(use_gpu=True) (or, do nothing); (2) if we want it to be run on the CPU (and not the GPU) in both machines (because, say, the test won't run on GPU) then we would use test_util.device(use_gpu=False); (3) if we want the test to fail if we try to run it on a machine with no GPU, then we would use test_util.force_gpu().

The question I have is: how do I develop and use these tests locally without having to add and remove the ability for them to be run on the different devices. In other words, I can't then as confidently submit what I have confirmed to be good code. I guess I can set CUDA_VISIBLE_DEVICES=-1 and run the test that way to check the non-GPU testing functionality. The problem with this is that I then have to pass the environment variable in on the bazel build, which is going to lead to a complete rebuild. So it would make development take forever. Alternatively, I would need two development machines and then I would need access to the code I'm developing from both machines at once (e.g. through a mounted drive). I already have to do a rebuild to check for the XLA case, but I usually do that at the very end of development and I'm usually not particularly developing to run on XLA (it just has to not fail).

I see a lot of tests that explicitly test on both CPU and GPU, and I written a lot of tests like this myself. In the above code, I have tried to optimize a little so that only the CPU case gets run on a machine with no GPU. Many tests, including the existing xent opt tests, will run the same case twice on a machine with no GPU.

Reply:
I'm aware that all the tests get run on machines with and without GPUs and that, therefore, the tests can be written so that each case will run on a GPU if a GPU is present and on a CPU if a GPU is not present. The reason I write tests like this, and (I presume) why others do (including the existing op test), is that it simplifies the development process: I can develop for both CPU and GPU on one machine with one build-test container.

Should I change my development practice to open both a GPU container and a non-GPU container on my machine, with the source code mounted into both, and then execute the build/run in both containers in parallel? The development procedure is already very cumbersome and slow; this would double the amount build/re-build time and things to keep track of.

Reply:
The next commit will remove this loop. It makes sense to remove it after development, when it is no longer needed.

Reply:
The next commit also removes all the extra cycles that were in the pre-existing test code to run test cases on both the CPU and GPU on a machine with a GPU (and which ran the same cases twice on machines without GPUs).

Reply:
You can run with `CUDA_VISIBLE_DEVICES=` (setting the env var to an empty string) if you want to test without a GPU. Admittedly, this is a pain since you cannot run both with and without the GPU in a single `bazel test` command.

/CC @sanjoy would it be feasible to have the non-GPU versions of test run with `CUDA_VISIBLE_DEVICES=` automatically?

Reply:
> Sorry for long periods between reviews, I had taken some time off.

No problem, @reedwm. I hope you had a relaxing and enjoyable break.